Refactor rule parsing

radumereuta opened the pull request:
Part 1 in fixing: #1867
Timings after changes (master):
make -j 32 regression-new:
memory usage: max 28.5GB (25.3GB)
time: 7m43 (8m47)

make evm-semantics:
make build - after changes:
time 1m15 (1m12)

kompile on an empty file:
time: 3.6s (3.9s)

make -j 25 pl-tutorial
master: 7m55, 23GB max
after changes: 7m50, 28GB max

There is a significant time improvement when running multiple tests.
When running all tests, there is a spike in memory usage at the beginning when I'm running 32 instances of kompile which all try to parse at the same time. I think this has increased because now I parse all bubbles at the same time. Before it was first all claims, then all rules, then all contexts, and finally all aliases.

I also had to fix a problem with `FlatModule` because it triggered some duplication errors. Before it was building lots of duplications and ended up being quadratic instead of linear complexity.

Here is the output of kompile on a test file with one rule and a configuration:
```
Outer parsing [132 modules]                                  =  0.639s
New scanner:TEST$SYNTAX-CONFIG-CELLS                         =  0.135s
New scanner:TEST-CONFIG-CELLS                                =  0.145s
New scanner:STDOUT-STREAM$SYNTAX-CONFIG-CELLS                =  0.215s
New scanner:STDOUT-STREAM-CONFIG-CELLS                       =  0.221s
New scanner:STDIN-STREAM$SYNTAX-CONFIG-CELLS                 =  0.218s
New scanner:STDIN-STREAM-CONFIG-CELLS                        =  0.227s
Parse configurations [6/6 declarations]                      =  1.650s
New scanner:TEST-RULE-CELLS                                  =  0.144s
New scanner:STDIN-STREAM-RULE-CELLS                          =  0.216s
New scanner:STDOUT-STREAM-RULE-CELLS                         =  0.231s
Parse rules [97/97 rules]                                    =  0.827s
```
second time:
```
Outer parsing [132 modules]                                  =  0.596s
New scanner:TEST$SYNTAX-CONFIG-CELLS                         =  0.131s
New scanner:TEST-CONFIG-CELLS                                =  0.153s
New scanner:STDOUT-STREAM$SYNTAX-CONFIG-CELLS                =  0.208s
New scanner:STDOUT-STREAM-CONFIG-CELLS                       =  0.217s
New scanner:STDIN-STREAM$SYNTAX-CONFIG-CELLS                 =  0.211s
New scanner:STDIN-STREAM-CONFIG-CELLS                        =  0.231s
Parse configurations [0/6 declarations]                      =  1.504s
Parse rules [0/97 rules]                                     =  0.075s
```
The diff was getting pretty big so I'm limiting the changes to rules only.
I'm going to start working on the configuration after this gets merged. I will deal with code duplication at the same time.

dwightguth commented on kore/src/main/scala/org/kframework/kore/transformers.scala:
```diff
@@ -0,0 +1,108 @@
+
+  def apply(k: KSequence): T = k.items.asScala.map(apply).fold(unit)(merge)
+
+  def apply(k: InjectedKLabel): T = unit
```
I do not see how this visits a klabel variable

dwightguth commented on kore/src/main/scala/org/kframework/kore/transformers.scala:
```diff
@@ -0,0 +1,108 @@
+trait FoldK[T] extends KTransformer[T] {
+
+  def apply(k: KApply): T = merge(
+    k.klabel match { case v: InjectedKLabel => apply(v); case _ => unit },
```
or this

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/checks/CheckRHSVariables.java:
```diff
@@ -0,0 +1,163 @@
+            public Set<KVariable> apply(KVariable v) {
+                return Set(v);
+            }
+        }.apply(KORE.KApply(KORE.KLabel(KLabels.ML_AND), KORE.KApply(KORE.KLabel(KLabels.ML_AND), k, requires), ensures));
```
I also do not see how this correctly treats variables in the requires clause as not part of the LHS.

cos replied:
this is the visitor for the RHS. The better name for it might the `otherVariables`

dwightguth replied:
oh, I see

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/checks/CheckRHSVariables.java:
```diff
@@ -0,0 +1,163 @@
+        }.apply(KORE.KApply(KORE.KLabel(KLabels.ML_AND), KORE.KApply(KORE.KLabel(KLabels.ML_AND), k, requires), ensures));
+
+        stream(rhsVars).forEach(v -> {
+            if (v.equals(ResolveAnonVar.ANON_VAR)
```
this also does not seem to correctly handle the case where an anonymous variable exists in part of a term that is both LHS and RHS.

cos replied:
correct. simply add `&& !lhsVars.contains(v)`

cos commented:
Actually, reverted those commits. I don't think `InjectedKlabel` should ever have a variable inside.

dwightguth commented:
Not true. In fact, we already have definitions where this will occur. Consider:
```
rule Lbl:KLabel(.KList) => .K
       requires #klabel(Lbl) in setOfKLabels
```

not being able to do this severely restricts the ability to do anything useful with klabel variables.

cos commented:
We went back and forth on this last year and I remember the conclusion was that they will get auto-injected. But I might be wrong. In any case, it's very easy to make the visitors check for it.

dwightguth commented:
Can I close this? You are welcome to refactor the check however you like as long as it  doesn't break the build for the C semantics.

cos commented:
Thanks! I'll integrate same of it in the next few days. Waiting for #1886. Leave it open so I don't forget.

cos commented:
might get back to it at some point

radumereuta committed:
Refactor rule parsing

Create a single function that deals with creating scanners and parsers to get more control over dependencies.

radumereuta committed:
Cleanup code.

radumereuta committed:
Extract upConfig

radumereuta committed:
Small fix

ehildenb commented on kernel/src/main/java/org/kframework/kompile/DefinitionParsing.java:
```diff
@@ -310,25 +300,14 @@ private Module resolveConfigBubbles(Definition def, Module inputModule, RuleGram
-                            throw KEMException.compilerError("Illegal configuration with requires clause detected.", configContents);
-                        }
-                    })
+                    .map(this::upConfig)
```
This change can likely be pulled out into a pure refactoring PR.

ehildenb commented on kernel/src/main/java/org/kframework/kprove/ProofDefinitionBuilder.java:
```diff
@@ -68,7 +68,11 @@ public ProofDefinitionBuilder(CompiledDefinition compiledDefinition, Backend bac
+        //Map<String, Module> modulesMap = modules.stream().collect(Collectors.toMap(Module::name, m -> m));
+        Map<String, List<Module>> group = modules.stream().collect(Collectors.groupingBy(Module::name));
+        Map<String, Module> modulesMap = group.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, x -> x.getValue().get(0)));
+
```
Can this change also come out in a refactoring PR?

ehildenb commented on kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java:
```diff
@@ -77,7 +78,7 @@ public ParseInModule(Module seedModule, Module extensionModule, Module disambMod
             try {
-                timing = new BufferedWriter(new FileWriter(files.resolveKompiled("timing" + Thread.currentThread().getId() + ".log"), true));
+                timing = new BufferedWriter(new FileWriter(files.resolveTemp("timing" + Thread.currentThread().getId() + ".log"), true));
             } catch (IOException e) {
```
Pure refactoring.

radumereuta committed:
Merge branch 'master' into refactorRuleParsing

radumereuta committed:
Undo refactoring changes

radumereuta committed:
Merge branch 'master' into refactorRuleParsing

radumereuta committed:
Fix indentation

radumereuta committed:
More reverts

radumereuta committed:
Eliminate config Bubbles in configParser

Simplifies further steps

radumereuta committed:
Fix a small bug

radumereuta committed:
Refactor FlatModule creation

radumereuta committed:
Refactor parser creation

radumereuta committed:
More code cleanup

radumereuta committed:
Merge branch 'master' into refactorRuleParsing

radumereuta committed:
Revert unnecessary changes

radumereuta committed:
Force order in which scanners to choose

Also fix a failing test

radumereuta committed:
Order modules to preserve error message for circular imports

radumereuta committed:
Small cleanup

radumereuta committed:
Make output synchronous

radumereuta committed:
Use test branch of the haskell backend.

ehildenb commented on k-distribution/pom.xml:
```diff
@@ -122,6 +122,7 @@
               <target>
                 <exec dir="${ktestDir}" executable="make" failonerror="true">
                   <arg value="-j12" />
+                  <arg value="--output-sync" />
```
```suggestion
                  <arg value="--output-sync=recurse" />
```

https://www.gnu.org/software/make/manual/html_node/Parallel-Output.html

ehildenb replied:
You'll want to change the Jenkinsfile where in calls `make` in the tutorial with teh same update.

ehildenb commented:
Subsumed by #1914 (among others)
